### PR TITLE
#357: phaenonet +

### PR DESCRIPTION
**Pull request by fabianpage**

### CODE REVIEW

**Comment by fabianpage** [comment]
deployed at https://www.youtube.com/watch?v=gbhfK5WPE7A

**Comment by fabianpage** [comment]
`test regression on map markers for 2018:` works on localhost

**Comment by pgoellnitz** [comment]
Is it possible to change the style how the checkboxes are 'checked' in the graph? It would be clearer if they would show a checkmark.

**Comment by pgoellnitz** [comment]
I was missing the script go gererate the Phaenonet+ icons. Could you also please add a line how to do it in the Readme?

**Review by pgoellnitz — CHANGES_REQUESTED** [review]
Looks great, some minor stuff. Let me know if the split of the header component is feasible.

File: src/app/map/map-overview.component.html [hunk 1]
```diff
@@ -17,10 +17,11 @@
           <div class="map-overview__info-window-text">
             <div class="map-overview__info-window-label">
               {{ d.species_name | translate }}
+              <app-sensors-badge [sensorLiveData]="d.sensor"></app-sensors-badge>
             </div>
             <div class="map-overview__info-window-value">{{ d.individual_name }}</div>
             <div class="map-overview__info-window-label" *ngIf="d.phenophase_name">
-              {{ 'Letzte Beobachtung' | translate }},
+              {{ 'Letzte Beobachtung' | translate }}:
```
[the remark this answers is not in the document]
Comment (fabianpage) [reply src/app/map/map-overview.component.html:24]: The `,` and the `:` are outside of the translated string.

File: src/app/sensors/sensors.service.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+  getSensorData(individual_id: string): Observable<DailySensorData[]> {
+    return this.afs
+      .collection<{ data: { [name: string]: Sensors }; year: number }>(this.collectionName)
+      .doc(individual_id)
+      .valueChanges()
+      .pipe(
+        map(sensors => {
+          const data = sensors.data;
+          const keys = Object.keys(data).filter(k => k != 'year');
```
[the remark this answers is not in the document]
Comment (fabianpage) [reply src/app/sensors/sensors.service.ts:24]: 👍  There was a sensor field at some point but that was probably an error

**Comment by fabianpage** [comment]
Hallo Philippe 

Ich mache mich die Tage mal noch ans aufsplitten der Header Componente, dies hat hetute nicht mehr gerreicht. 

Liebe Grüsse 
Fabian 

-- 
Fabian Page 
***@***.*** 
+41 31 328 33 60 

Tegonal Genossenschaft 
Wasserwerkgasse 2 
3011 Bern, Switzerland 
http://tegonal.com 

-- 
http://twitter.com/tegonal 
http://github.com/tegonal 


Von: "Philippe Göllnitz" ***@***.***> 
An: "globe-swiss/phaenonet-client" ***@***.***> 
CC: "fabian page" ***@***.***>, "Author" ***@***.***> 
Gesendet: Montag, 14. November 2022 08:13:06 
Betreff: Re: [globe-swiss/phaenonet-client] #357: phaenonet + (PR #379) 





@pgoellnitz requested changes on this pull request. 

Looks great, some minor stuff. Let me know if the split of the header component is feasible. 


In [ https://github.com/globe-swiss/phaenonet-client/pull/379#discussion_r1021113817 | src/app/individual/individual-header/individual-header.component.ts ] : 
>  import { MasterdataService } from 'src/app/masterdata/masterdata.service';
 import { Individual } from '../individual';
 import { IndividualService } from '../individual.service';
 import { GeoposService } from './geopos.service';
+import * as d3Axis from 'd3-axis'; 


I would really prefer if the components for map and graph would be separated. 


In [ https://github.com/globe-swiss/phaenonet-client/pull/379#discussion_r1021115569 | src/app/individual/individual.ts ] : 
> @@ -31,12 +33,29 @@ export interface Individual {
   image_urls: string[];
 }
 
+export interface SensorLiveData {
+  ah: number; // air humidity
+  at: number; // air temperature
+  sh: number; // soil humidity
+  st: number; // soil temperature
+  ts: { seconds: number; nanoseconds: number }; // last update
+}
+
+export function isMapindividual(individual: MapIndividual | Individual): individual is MapIndividual {
+  return (individual as MapIndividual).has_sensor !== undefined; 


I believe this should be evaluated to 'true' as has_sensor is boolean. 


In [ https://github.com/globe-swiss/phaenonet-client/pull/379#discussion_r1021123196 | src/app/sensors/sensors.service.ts ] : 
> +
+  getSensorData(individual_id: string): Observable<DailySensorData[]> {
+    return this.afs
+      .collection<{ data: { [name: string]: Sensors }; year: number }>(this.collectionName)
+      .doc(individual_id)
+      .valueChanges()
+      .pipe(
+        map(sensors => {
+          const data = sensors.data;
+          const keys = Object.keys(data).filter(k => k != 'year');
+
+          const v = keys.sort().map(key => ({ day: key, ...data[key] }));
+          return v.map(this.#toDailySensorData);
+        })
+      )
+      .pipe(tap(x => this.fds.addRead(`${this.collectionName} (listByIds)`, x.length))); 


listByIds -> getSensorData 


In [ https://github.com/globe-swiss/phaenonet-client/pull/379#discussion_r1021125292 | src/app/sensors/sensors.service.ts ] : 
> +
***@***.***()
+export class SensorsService extends BaseResourceService<Sensors> {
+  constructor(alertService: AlertService, protected afs: AngularFirestore, protected fds: FirestoreDebugService) {
+    super(alertService, afs, 'sensors', fds);
+  }
+
+  getSensorData(individual_id: string): Observable<DailySensorData[]> {
+    return this.afs
+      .collection<{ data: { [name: string]: Sensors }; year: number }>(this.collectionName)
+      .doc(individual_id)
+      .valueChanges()
+      .pipe(
+        map(sensors => {
+          const data = sensors.data;
+          const keys = Object.keys(data).filter(k => k != 'year'); 


Do we need this filter criteria? There should be no 'year' field in the 'data' map. 


In [ https://github.com/globe-swiss/phaenonet-client/pull/379#discussion_r1021131166 | src/app/map/map-overview.component.html ] : 
>              </div>
             <div class="map-overview__info-window-value">{{ d.individual_name }}</div>
             <div class="map-overview__info-window-label" *ngIf="d.phenophase_name">
-              {{ 'Letzte Beobachtung' | translate }},
+              {{ 'Letzte Beobachtung' | translate }}: 


If we change it here we also need to change the translations.

**Review by pgoellnitz — APPROVED** [review]
lgtm, thanks a bunch!

**Pull request merged**